src/corefx/System.Globalization.Native/casing.cpp

```diff
@@ -0,0 +1,65 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+#include <assert.h>
+#include <stdint.h>
+#include <unicode/uchar.h>
+#include <unicode/utf16.h>
+
+/*
+Function:
+ChangeCase
+
+Preforms upper or lower casing of a string into a new buffer, preforming special casing for turkish, if needed.
+*/
+extern "C" void ChangeCase(const UChar* lpSrc, int32_t cwSrcLength, UChar* lpDst, int32_t cwDstLength, int32_t bToUpper, int32_t bTurkishCasing)
+{
+    int32_t srcIdx = 0;
+    int32_t dstIdx = 0;
+
+    UBool isError = FALSE;
+
+    while (srcIdx < cwSrcLength)
+    {
+        UChar32 srcCodepoint;
+        UChar32 dstCodepoint;
+
+        // Decode the next one or two UTF-16 code units into a codepoint and update srcIdx to point to the next UTF-16 code unit to decode.
+        U16_NEXT(lpSrc, srcIdx, cwSrcLength, srcCodepoint);
+
+        if (bToUpper)
+        {
+            if (!bTurkishCasing)
+            {
+                dstCodepoint = u_toupper(srcCodepoint);
+            }
+            else
+            {
+                // In turkish casing, LATIN SMALL LETTER I (U+0069) upper cases to LATIN CAPITAL LETTER I WITH DOT ABOVE (U+0130).
+                dstCodepoint = ((srcCodepoint == (UChar32)0x0069) ? (UChar32)0x0130 : u_toupper(srcCodepoint));
+            }
+        }
+        else
+        {
+            if (!bTurkishCasing)
+            {
+                dstCodepoint = u_tolower(srcCodepoint);
+            }
+            else
+            {
+                // In turkish casing, LATIN CAPITAL LETTER I (U+0049) lower cases to LATIN SMALL LETTER DOTLESS I (U+0131).
+                dstCodepoint = ((srcCodepoint == (UChar32)0x0049) ? (UChar32)0x0131 : u_tolower(srcCodepoint));
+            }
+        }
```

Comment:
Would it be better / more efficient to have four loops that we choose between based on bToUpper and bTurkishCasing, rather than having a single loop and having to do two branches per iteration?  Or is the compiler able to optimize this accordingly?